src/katgpucbf/utils.py

```diff
@@ -70,36 +70,48 @@ class DeviceStatus(Enum):
 
     This takes its value from the worst status of its target set of sensors, so
     it's quick to identify if there's something wrong, or if everything is good.
     """
 
-    def __init__(self, target: aiokatcp.SensorSet):
-        super().__init__(target=target, sensor_type=DeviceStatus, name="device-status")
+    def __init__(self, target: aiokatcp.SensorSet) -> None:
+        super().__init__(
+            target=target, sensor_type=DeviceStatus, name="device-status", description="Overall engine health"
+        )
 
     def update_aggregate(
         self,
-        updated_sensor: Optional[aiokatcp.Sensor[_T]],
-        reading: Optional[aiokatcp.Reading[_T]],
-        old_reading: Optional[aiokatcp.Reading[_T]],
-    ) -> aiokatcp.Reading[DeviceStatus]:  # noqa: D102
-        # For device status it's far simpler just to re-calculate everything
-        # each time, than to try and maintain state.
-        worst_status: aiokatcp.Sensor.Status = aiokatcp.Sensor.Status.UNKNOWN
+        updated_sensor: aiokatcp.Sensor[_T] | None,
+        reading: aiokatcp.Reading[_T] | None,
+        old_reading: aiokatcp.Reading[_T] | None,
+    ) -> aiokatcp.Reading[DeviceStatus] | None:  # noqa: D102
+
+        if reading is not None and old_reading is not None and reading.status == old_reading.status:
+            return None  # Sensor didn't change state, so no change in overall device status
+        # Otherwise, we have to recalculate.
+        worst_status: aiokatcp.Sensor.Status = aiokatcp.Sensor.Status.NOMINAL
         for sensor in self.target.values():
             if self.filter_aggregate(sensor):
                 worst_status = max(worst_status, sensor.status)
 
+        if reading is not None and old_reading is not None:  # i.e. an update
+            timestamp = sensor.timestamp
+        else:  # i.e. an addition or removal
+            timestamp = time.time()
+        # max in order to prevent time from appearing to move backwards
+        # if sensors' timestamps come from different places
+        timestamp = max(timestamp, self.timestamp)
+
         if worst_status == aiokatcp.Sensor.Status.NOMINAL:
-            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.NOMINAL, DeviceStatus.OK)
-        if worst_status == aiokatcp.Sensor.Status.WARN:
-            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.WARN, DeviceStatus.DEGRADED)
-        if worst_status == aiokatcp.Sensor.Status.UNKNOWN:  # Do we need to take this situation into account?
-            return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.UNKNOWN, DeviceStatus.UNKNOWN)
-        return aiokatcp.Reading(sensor.timestamp, aiokatcp.Sensor.Status.ERROR, DeviceStatus.FAIL)
+            agg_value = DeviceStatus.OK
+        elif worst_status == aiokatcp.Sensor.Status.WARN:
+            agg_value = DeviceStatus.DEGRADED
+        else:
+            agg_value = DeviceStatus.FAIL
```

Comment:
Since we're using Python 3.10 this could be a reasonable place to apply the new pattern matching :-)

Reply:
I did think about that. For quite a while. But I ultimately didn't feel like arguing over whether it's a feature that should be used or avoided.

I do think the way I've phrased it now is more clear than what was there previously though.

Reply:
Yes, it's definitely clear now. If there were more branches I'd suggest with pattern-matching or a lookup dictionary (using `.get` to provide a default for everything else), but with the current number of branches it's perfectly readable so I'm happy to leave it.